flink/src/test/java/org/apache/iceberg/actions/TestRewriteDataFilesAction.java

```diff
@@ -0,0 +1,211 @@
+    List<DataFile> dataFiles = Lists.newArrayList(CloseableIterable.transform(tasks, FileScanTask::file));
+    Assert.assertEquals("Should have 4 data files before rewrite", 4, dataFiles.size());
+
+    RewriteDataFilesActionResult result =
+        Actions.forTable(icebergTablePartitioned)
+            .rewriteDataFiles()
+            .execute();
+
+    Assert.assertEquals("Action should rewrite 4 data files", 4, result.deletedDataFiles().size());
+    Assert.assertEquals("Action should add 2 data file", 2, result.addedDataFiles().size());
+
+    icebergTablePartitioned.refresh();
+
+    CloseableIterable<FileScanTask> tasks1 = icebergTablePartitioned.newScan().planFiles();
+    List<DataFile> dataFiles1 = Lists.newArrayList(CloseableIterable.transform(tasks1, FileScanTask::file));
+    Assert.assertEquals("Should have 2 data files after rewrite", 2, dataFiles1.size());
+
+    // Assert the table records as expected.
+    SimpleDataUtil.assertTableRecords(icebergTablePartitioned, Lists.newArrayList(
+        SimpleDataUtil.createRecord(1, "hello"),
+        SimpleDataUtil.createRecord(2, "hello"),
+        SimpleDataUtil.createRecord(3, "world"),
+        SimpleDataUtil.createRecord(4, "world")
+    ));
+  }
+
+
+  @Test
+  public void testRewriteDataFilesWithFilter() throws Exception {
+    sql("INSERT INTO %s SELECT 1, 'hello' ", TABLE_NAME_PARTIITONED);
+    sql("INSERT INTO %s SELECT 1, 'hello' ", TABLE_NAME_PARTIITONED);
+    sql("INSERT INTO %s SELECT 1, 'world' ", TABLE_NAME_PARTIITONED);
+    sql("INSERT INTO %s SELECT 2, 'world' ", TABLE_NAME_PARTIITONED);
+    sql("INSERT INTO %s SELECT 3, 'world' ", TABLE_NAME_PARTIITONED);
+
+    icebergTablePartitioned.refresh();
+
+    CloseableIterable<FileScanTask> tasks = icebergTablePartitioned.newScan().planFiles();
+    List<DataFile> dataFiles = Lists.newArrayList(CloseableIterable.transform(tasks, FileScanTask::file));
+    Assert.assertEquals("Should have 5 data files before rewrite", 5, dataFiles.size());
+
+    RewriteDataFilesActionResult result =
+        Actions.forTable(icebergTablePartitioned)
+            .rewriteDataFiles()
+            .filter(Expressions.equal("id", 1))
+            .filter(Expressions.startsWith("data", "he"))
+            .execute();
+
+    Assert.assertEquals("Action should rewrite 2 data files", 2, result.deletedDataFiles().size());
+    Assert.assertEquals("Action should add 1 data file", 1, result.addedDataFiles().size());
+
+    icebergTablePartitioned.refresh();
+
+    CloseableIterable<FileScanTask> tasks1 = icebergTablePartitioned.newScan().planFiles();
+    List<DataFile> dataFiles1 = Lists.newArrayList(CloseableIterable.transform(tasks1, FileScanTask::file));
+    Assert.assertEquals("Should have 4 data files after rewrite", 4, dataFiles1.size());
+
+    // Assert the table records as expected.
+    SimpleDataUtil.assertTableRecords(icebergTablePartitioned, Lists.newArrayList(
```

Comment:
I think we may need to refactor the `SimpleDataUtil.assertTableRecords`, because it will use `Set` to check the expected records, that means it will de-duplicate the same records automatically, and won't identify the case that have two `<1, 'hello'>` records.